File: actito_flutter/ios/Classes/Events/ActitoEventStream.swift
//
//  ActitoEventStream.swift
//  actito
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation

class ActitoEventStream: NSObject, FlutterStreamHandler {

    let eventType: ActitoEventType
    let name: String

    private var eventSink: FlutterEventSink?
    private var pendingEvents: [ActitoEvent] = []

    init(eventType: ActitoEventType) {
        self.eventType = eventType
        self.name = "com.actito.flutter/events/\(eventType.rawValue)"
    }

    func send(_ event: ActitoEvent) {
        if let sink = self.eventSink {
            sink(event.payload)
        } else {
            pendingEvents.append(event)
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        if self.eventSink != nil {
            self.pendingEvents.forEach { send($0) }
            self.pendingEvents.removeAll()
        }

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: actito_inbox/android/src/main/kotlin/com/actito/inbox/flutter/events/ActitoEventStream.kt
package com.actito.inbox.flutter.events

import io.flutter.plugin.common.EventChannel

internal class ActitoEventStream(type: ActitoEvent.Type) : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null
    private val pendingEvents = mutableListOf<ActitoEvent>()

    val name = "com.actito.inbox.flutter/events/${type.id}"

    fun emit(event: ActitoEvent) {
        val eventSink = this.eventSink

        if (eventSink == null) {
            pendingEvents.add(event)
        } else {
            eventSink.success(event.payload)
        }
    }

    override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink

        if (eventSink != null) {
            pendingEvents.forEach(::emit)
            pendingEvents.clear()
        }
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }
}
